### ISSUE TITLE
Remove code generation from the Terraform bridge

### ISSUE THREAD

**Issue by scottyw** [lyraproj/lyra#66]
The Terraform bridge uses code generation to create a Go struct and handler for every supported AWS type, hundreds in total. At runtime our provider interface deals with meta-level descriptions of types and functions which are mapped to those concrete types which are in turn passed to the provider CRUD functions. To integrate with Terraform content, our provider uses one of many almost-identical bridging provider implementations. These convert the concrete types back into a meta-level descriptions (i.e. `*terraform.ResourceConfig`) before calling the TF provider.

In principle we can instead generate all Lyra type information directly from the TF schema at runtime, without the need to use a code generator to produce intermediate concrete types. Similarly a single provider implementation could handle all possible TF types.

This issue captures design and implementation of this approach.

**Comment by scottyw** [comment]
I'd like to try adding some new functionality to `service.Server`. Today we register type X and associate it with struct type Y. When the handle is called, the server will deserialize incoming `eval.Value`s of type X into instances of Y and call the CRUD functions with that new instance.

I'd like to extend `Server` to support custom serialization/deserialization of registered types. So when an `eval.Value` of type X arrives it would be deserialized using a custom deserializer before being passed to the handler. I think that would allow us to build a single "Terraform handler" that accepts a single concrete type (perhaps `terraform.ResourceConfig` or perhaps just `map[string]interface{}`) and associate that handler with all registered TF types.

@thallgren @markfuller What do you reckon?

**Comment by scottyw** [comment]
We'll probably use PuppetObject as type X above i.e. it will be possible to write a CRUD handler in terms of PuppetObject to support this dynamic use case. An instance of the eval.PuppetObject can be created using the eval.New() and passing a context, the resource type, and a hash as arguments.

**Comment by scottyw** [comment]
An update: currently we generate the Go structs for all resource types plus a per-resource provider. We no longer generate functions to map between Go structs and TF map[string]interface{}, instead marshaling and unmarshaling reflectively.

The final step is to replace the per-resource provider with a single PuppetObject based provider. Each TF resource could then be registered directly based on the schema rather than derived from a Go struct.

**Comment by ahpook** [comment]
@scottyw do i understand correctly than when this is done, we wouldn't need to know beforehand about a particular terraform provider in order to make use of it in lyra? this seems like it'd be a huge win for the content ecosystem.

**Comment by ahpook** [comment]
also you may have seen this already, but here's the pulumi - terraform bridge code: https://github.com/pulumi/pulumi-terraform

**Comment by scottyw** [comment]
@ahpook We leaned heavily on the Pulumi bridge and managed to reuse some code here: https://github.com/lyraproj/lyra/blob/30048ff4483dec14b6b8379d6ed84bcb635656e8/pkg/bridge/bridge.go

I don't think we can get to the point there we're fully dynamic, such that don't need to know in advance about a provider to use it. I think of that as binary compatibility. Our bridge depends on linking to source code because we hook in a bit lower in the stack. 

The hope with this ticket is that we can remove the code generation so that adding support for a new provider becomes almost a one-liner. Where today we generate the code like this:

```go
bridge.Generate(aws.Provider().(*schema.Provider), "TerraformAws", "cmd/goplugin-tf-aws/generated/generated.go")
```

you could imagine something like:

```go
bridge.Serve(aws.Provider().(*schema.Provider), "TerraformAws")
```

Stick that in a `main` and you have your provider without the need to generate and keep any supporting source code.

**Comment by scottyw** [comment]
I'm dragging my feet around this issue a bit because the experience today is fairly decent and I reckon it might be time-consuming enough to implement.

**Comment by ahpook** [comment]
No worries, thanks for the clarification.

**Comment by thallgren** [comment]
Closed by lyraproj/terraform-bridge#2.
